### PR TITLE
 [ENH] Online learning utilities

### PR DESCRIPTION
**Pull request by HugoDelatte**
* Add online learning utilities (`online_score`, `online_predict`, `OnlineGridSearch` and `OnlineRandomizedSearch`)
* Add covariance forecast evaluation utilities
* Add partial_fit support for `EmpiricalPrior`
* Add partial_fit support for `MeanRisk`
* Update User Guide with online learning section
* Update Examples with online learning section

### CODE REVIEW

File: examples/online_learning/plot_1_online_covariance_forecast_evaluation.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+# calibrated, with a target of 1.0. We expect the regime-adjusted model to remain
+# closer to 1.0, especially during stress periods.
+comparison.plot_calibration(diagnostics=["bias"])
+
+# %%
+# QLIKE Loss
+# ==========
+# Let's now plot the QLIKE loss. It compares the forecast portfolio variance
+# with the realized sum of squared portfolio returns over the evaluation
+# window, so lower values indicate better portfolio variance forecasts.
```
Comment (FarrellDaniel) [inline examples/online_learning/plot_1_online_covariance_forecast_evaluation.py]: reads slightly clunky - would say "with lower values indicating" instead of "so"

File: examples/online_learning/plot_2_online_hyperparameter_tuning.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+setting using :class:`~skfolio.model_selection.OnlineGridSearch` and
+:class:`~skfolio.model_selection.OnlineRandomizedSearch`.
+
+The online approach is equivalent to combining scikit-learn's
+:class:`~sklearn.model_selection.GridSearchCV`
+(or :class:`~sklearn.model_selection.RandomizedSearchCV`) with
+:class:`~skfolio.model_selection.WalkForward` using `expand_train=True`, but instead of
+refitting every candidate from scratch at each split, it calls `partial_fit` to
+incrementally update the estimator, which is significantly faster for estimators that
+support it.
```
Comment (FarrellDaniel) [inline examples/online_learning/plot_2_online_hyperparameter_tuning.py]: Sentence is a bit bloated and 'it' doesn't read so well at the end. Perhaps small rewrite to: ..."but instead of
refitting every candidate from scratch at each split, it calls `partial_fit` to incrementally update the estimator. This is significantly faster for estimators that support this method."

File: examples/online_learning/plot_3_online_portfolio_optimization_evaluation.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+<sphx_glr_auto_examples_online_learning_plot_2_online_hyperparameter_tuning.py>`,
+which tuned the covariance estimator in isolation, here we optimize the portfolio model
+end-to-end using a portfolio-level metric.
+
+The online approach is equivalent to combining scikit-learn's
+:class:`~sklearn.model_selection.GridSearchCV` with
+:class:`~skfolio.model_selection.WalkForward` using `expand_train=True`, but instead of
+refitting every candidate from scratch at each split, it calls `partial_fit` to
+incrementally update each estimator, which is significantly faster for estimators that
+support it.
```
Comment (FarrellDaniel) [inline examples/online_learning/plot_3_online_portfolio_optimization_evaluation.py]: same comment as above regarding full wording - _Sentence is a bit bloated and 'it' doesn't read so well at the end. Perhaps small rewrite to: ..."but instead of
refitting every candidate from scratch at each split, it calls partial_fit to incrementally update the estimator. This is significantly faster for estimators that support this method"_

File: src/skfolio/metrics/_covariance.py [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+    portfolio_weights : array-like of shape (n_assets,) or (n_portfolios, n_assets), optional
+        Portfolio weights. If `None` (default), inverse-volatility weights are used,
+        which neutralizes volatility dispersion so that high-volatility assets do not
+        dominate the diagnostic. If a 2D array is provided, each row defines a test
+        portfolio and the mean QLIKE across portfolios is returned.
+
+    Returns
+    -------
+    float
+        Mean portfolio QLIKE loss. Lower values are better.
```
Comment (FarrellDaniel) [inline src/skfolio/metrics/_covariance.py]: perhaps would be useful to qualify this "lower is better" with the fuller comment, "In expectation, the loss is minimized by the true conditional portfolio variance forecast."

File: src/skfolio/metrics/_covariance.py [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+    confidence_level: float,
+) -> float:
+    r"""Exceedance rate for chi-squared calibration statistics.
+
+    Computes the fraction of squared distances exceeding the upper `confidence_level`
+    chi-squared quantile.
+
+    The reference threshold assumes Gaussian standardized returns. In practice,
+    the rate is sensitive not only to covariance misspecification but also to
+    heavy tails, regime shifts, and non-Gaussian standardized returns.  It is
```
Comment (FarrellDaniel) [inline src/skfolio/metrics/_covariance.py]: double space after "non-Gaussian standardized returns. __It is..."

File: src/skfolio/metrics/_covariance.py [hunk 6]
```diff
@@ -0,0 +1,766 @@
...
+        Realized returns.
+
+    forecast_variance : array-like of shape (n_observations,)
+        Forecast variances for the same timestamps, expressed in squared
+        return units.
+
+    Returns
+    -------
+    float
+        Mean QLIKE loss. Lower values are better.
```
Comment (FarrellDaniel) [inline src/skfolio/metrics/_covariance.py]: same as for portfolio_variance_qlike_loss

File: src/skfolio/metrics/_covariance.py [hunk 7]
```diff
@@ -0,0 +1,766 @@
...
+    --------
+    portfolio_variance_qlike_loss : Multivariate QLIKE loss projected onto
+        portfolio weights.
+
+    References
+    ----------
+    .. [1] Patton, A. J. (2011). "Volatility forecast comparison using
+        imperfect volatility proxies." Journal of Econometrics, 160(1),
+        246-256.
+    """
```
Comment (FarrellDaniel) [inline src/skfolio/metrics/_covariance.py]: don't see where the reference is required in qlike_loss()

File: src/skfolio/model_selection/_online/_search.py [hunk 8]
```diff
@@ -0,0 +1,1040 @@
...
+
+    def score(self, X: npt.ArrayLike, y=None):
+        """Score using the best estimator found during search.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_observations, n_assets)
+            Price returns.
+
+        y : ignored
```
Comment (FarrellDaniel) [inline src/skfolio/model_selection/_online/_search.py]: Doesn't follow same convention of why you ignore and include (for scikit API compatibility)...but potentially that is intentional here.

File: src/skfolio/model_selection/_online/_search.py [hunk 9]
```diff
@@ -0,0 +1,1040 @@
...
+
+
+class OnlineRandomizedSearch(BaseOnlineSearch):
+    """Online randomized search on hyper parameters.
+
+    Each sampled parameter combination is evaluated by running a full online
+    walk-forward pass. Unlike :class:`OnlineGridSearch`, not all parameter values are
+    tried out but a fixed number of parameter settings is sampled from the specified
+    distributions. The number of parameter settings that are tried is given by `n_iter`.
+
```
Comment (FarrellDaniel) [inline src/skfolio/model_selection/_online/_search.py:643]: should be 'are' instead of 'is': 

not all parameters are tried out, but a fixed number of parameter settings are sampled from the specified distributions.

File: src/skfolio/model_selection/_online/_validation.py [hunk 10]
```diff
@@ -0,0 +1,911 @@
...
+    test_size: int = 1,
+    freq: str | pd.offsets.BaseOffset | None = None,
+    freq_offset: pd.offsets.BaseOffset | dt.timedelta | None = None,
+    previous: bool = False,
+    purged_size: int = 0,
+    reduce_test: bool = False,
+    params: dict | None = None,
+    portfolio_params: dict | None = None,
+) -> MultiPeriodPortfolio:
+    r"""Generate out-of-sample portfolios using online (incremental) learning.
```
Comment (FarrellDaniel) [inline src/skfolio/model_selection/_online/_validation.py]: Only definition so far (to my memory/search...) which states "(incremental)" --> perhaps better to keep uniform with the remaining phrasing which is just "online learning"?

File: src/skfolio/model_selection/_online/_validation.py [hunk 11]
```diff
@@ -0,0 +1,911 @@
...
+
+    Walks forward through the data, updating the estimator incrementally via
+    `partial_fit` and predicting on each subsequent test window. Unlike
+    :func:`~skfolio.model_selection.cross_val_predict`, which clones the estimator for
+    each fold, this function maintains a single stateful estimator that accumulates
+     knowledge over time.
```
Comment (FarrellDaniel) [inline src/skfolio/model_selection/_online/_validation.py]: leading space --> "_knowledge over time." vs. "knowledge over time."

File: src/skfolio/metrics/_covariance.py (hunk 5 above)
Comment (HugoDelatte) [reply src/skfolio/metrics/_covariance.py]: Sharp eye!

File: src/skfolio/model_selection/_covariance_forecast_evaluation.py [hunk 12]
```diff
@@ -0,0 +1,1562 @@
...
+          indicate underestimated risk, and values below `1.0` indicate overestimated
+          risk.When only one portfolio is evaluated, the `std` column is exactly that
+          portfolio's bias statistic. When multiple portfolios are evaluated,
+          portfolio-level diagnostics are first computed separately for each portfolio
+          and then aggregated by their median. In particular, the `std` column becomes
+          the median of the per-portfolio bias statistics. See also
+          :attr:`bias_statistic` and :meth:`bias_statistic_summary`.
+
+        * For QLIKE loss, there is no fixed numeric target. Accordingly,
+        `mad_from_target` is NaN and `target` is `"lower is better"`.
```
Comment (FarrellDaniel) [inline src/skfolio/model_selection/_covariance_forecast_evaluation.py]: missing indentation?

File: src/skfolio/model_selection/_covariance_forecast_evaluation.py [hunk 13]
```diff
@@ -0,0 +1,1562 @@
...
+            )
+        fig.update_yaxes(tickformat=".0%", rangemode="tozero")
+        return fig
+
+
+@dataclass(frozen=True, eq=False)
+class CovarianceForecastComparison:
+    r"""Side-by-side comparison of covariance forecast evaluations.
+
+    Aggregates multiple :class:`CovarianceForecastEvaluation` instances and  provides
```
Comment (FarrellDaniel) [inline src/skfolio/model_selection/_covariance_forecast_evaluation.py]: double space :  instances and__provides

File: src/skfolio/model_selection/_covariance_forecast_evaluation.py [hunk 14]
```diff
@@ -0,0 +1,1562 @@
...
+        """Iterate over (name, evaluation) pairs."""
+        return zip(self._names, self.evaluations, strict=True)
+
+    def summary(self) -> pd.DataFrame:
+        r"""Consolidated summary statistics for all estimators.
+
+        Returns a DataFrame with metrics as rows and a column-level
+        MultiIndex `(estimator, stat)` where stat is one of `mean`,
+        `median`, `std`, `p5`, `p95`, `mad_from_target`,
+        `target`.
```
Comment (FarrellDaniel) [inline src/skfolio/model_selection/_covariance_forecast_evaluation.py]: "'target'." should be able to fit on previous line?

File: src/skfolio/model_selection/_validation.py [hunk 15]
```diff
@@ -330,6 +305,123 @@ def _routing_enabled() -> bool:
...
+
+    params : dict, optional
+        Raw parameters from the caller.
+
+    owner : str
+        Name of the calling function, used in error messages.
+
+    callee : str
+        Estimator method that will receive the routed parameters. Use `"fit"` for batch
+        evaluation and `"partial_fit"` for online (incremental) evaluation.
```
Comment (FarrellDaniel) [inline src/skfolio/model_selection/_validation.py]: second instance of "online (incremental)" and here refers to "evaluation" as opposed to "learning" - not necessarily wrong just not 100% consistent naming

File: src/skfolio/model_selection/_validation.py [hunk 16]
```diff
@@ -330,6 +305,123 @@ def _routing_enabled() -> bool:
...
+            unrequested_params = sorted(e.unrequested_params)
+            raise ske.UnsetMetadataPassedError(
+                message=(
+                    f"{unrequested_params} are passed to `{owner}` but are"
+                    " not explicitly set as requested or not requested for"
+                    f" {owner}'s estimator: "
+                    f"{estimator.__class__.__name__}. Call"
+                    f" `.{request_method}({{metadata}}=True)` on the estimator"
+                    f" for each metadata in {unrequested_params} that you want"
+                    " to use and `metadata=False` for not using it. See the"
```
Comment (FarrellDaniel) [inline src/skfolio/model_selection/_validation.py]: "for not using it" --> "if you are not using it" or "to not use it" or another formulation

**Review by FarrellDaniel — DISMISSED** [review]
Aside small wording suggestions/whitespace typos, fine to merge.
